packages/trpc/server/routers/viewer/attributes/assignUserToAttribute.handler.ts

```diff
@@ -0,0 +1,239 @@
+import slugify from "@calcom/lib/slugify";
+import prisma from "@calcom/prisma";
+
+import { TRPCError } from "@trpc/server";
+
+import type { TrpcSessionUser } from "../../../trpc";
+import type { ZAssignUserToAttribute } from "./assignUserToAttribute.schema";
+
+type GetOptions = {
+  ctx: {
+    user: NonNullable<TrpcSessionUser>;
+  };
+  input: ZAssignUserToAttribute;
+};
+
+function isOrgAdminOrThrow(ctx: GetOptions["ctx"]) {
+  const org = ctx.user.organization;
+  if (!org.isOrgAdmin) {
+    throw new TRPCError({
+      code: "UNAUTHORIZED",
+      message: "You need to be an admin of the organization to use this feature",
+    });
+  }
+}
+
+const assignUserToAttributeHandler = async ({ input, ctx }: GetOptions) => {
+  const org = ctx.user.organization;
+
+  if (!org.id) {
+    throw new TRPCError({
+      code: "UNAUTHORIZED",
+      message: "You need to be apart of an organization to use this feature",
+    });
+  }
+
+  isOrgAdminOrThrow(ctx);
+
+  // TODO: We need to also empty the users assignemnts for IDs that are not in in this filteredAttributes list
+  // Filter out attributes that don't have a value or options set
+  const filteredAttributes = input.attributes.filter((attribute) => attribute.value || attribute.options);
+
+  // Ensure this organization can access these attributes and attribute options
+  const attributes = await prisma.attribute.findMany({
+    where: {
+      id: {
+        in: filteredAttributes.map((attribute) => attribute.id),
+      },
+      teamId: org.id,
+    },
+    select: {
+      id: true,
+      type: true,
+      options: true,
+    },
+  });
+
+  if (attributes.length !== filteredAttributes.length) {
+    throw new TRPCError({
+      code: "UNAUTHORIZED",
+      message: "You do not have access to these attributes",
+    });
+  }
+
+  const arrayOfAttributeOptionIds = attributes.flatMap(
+    (attribute) => attribute.options?.map((option) => option.id) || []
+  );
+
+  const attributeOptionIds = Array.from(new Set(arrayOfAttributeOptionIds));
+
+  const attributeOptions = await prisma.attributeOption.findMany({
+    where: {
+      id: {
+        in: attributeOptionIds,
+      },
+      attribute: {
+        teamId: org.id,
+      },
+    },
+    select: {
+      id: true,
+      value: true,
+      slug: true,
+    },
+  });
+
+  if (attributeOptions.length !== attributeOptionIds.length) {
+    throw new TRPCError({
+      code: "UNAUTHORIZED",
+      message: "You do not have access to these attribute options",
+    });
+  }
+
+  const membership = await prisma.membership.findFirst({
+    where: {
+      userId: input.userId,
+      teamId: org.id,
+    },
+  });
+
+  if (!membership) {
+    throw new TRPCError({
+      code: "UNAUTHORIZED",
+      message: "This user is not apart of your organization",
+    });
+  }
+
+  // const promises: Promise<{ id: string }>[] = [];
+
+  filteredAttributes.map(async (attribute) => {
```

Comment:
I noticed in this map we're making a lot of DB queries and writes. Instead could we make a singular DB call to get the attributes associated with the user and then a single `updateMany` call after iterating over `filteredAttributes`?

Reply:
I do agree with you here - but i battled and battled transactions - trying to do this in a cleaner way and everything i was trying was actually slower than this approach. 

If you have any ideas would love to try them

Reply:
I just took a stab at this again - we can't really update any as all options for example have different values and not the same.